crates/hir-def/src/resolver.rs

```diff
@@ -163,22 +169,33 @@ impl Resolver {
         self.resolve_module_path(db, path, BuiltinShadowMode::Module)
     }
 
-    pub fn resolve_path_in_type_ns(
-        &self,
-        db: &dyn DefDatabase,
-        path: &Path,
-    ) -> Option<(TypeNs, Option<usize>, Option<ImportOrExternCrate>)> {
+    pub fn resolve_path_in_type_ns<'a>(
+        &'a self,
+        db: &'a dyn DefDatabase,
+        path: &'a Path,
+    ) -> impl Iterator<Item = (ModuleOrTypeNs, Option<usize>, Option<ImportOrExternCrate>)> + 'a
+    {
         self.resolve_path_in_type_ns_with_prefix_info(db, path).map(
-            |(resolution, remaining_segments, import, _)| (resolution, remaining_segments, import),
+            move |(resolution, remaining_segments, import, _)| {
+                (resolution, remaining_segments, import)
+            },
         )
     }
 
-    pub fn resolve_path_in_type_ns_with_prefix_info(
-        &self,
-        db: &dyn DefDatabase,
-        path: &Path,
-    ) -> Option<(TypeNs, Option<usize>, Option<ImportOrExternCrate>, ResolvePathResultPrefixInfo)>
-    {
+    pub fn resolve_path_in_type_ns_with_prefix_info<'a>(
+        &'a self,
+        db: &'a dyn DefDatabase,
+        path: &'a Path,
+    ) -> Box<
```

Comment:
`Box<dyn Iterator>` optimizes very bad and also incurs an allocation, and resolver is very hot.

In general I don't think an iterator is the correct thing to return here. Resolution doesn't have "multiple options". It has a single option. If it resolves the a module, it resolves to it and to nothing else. So this should return a single value.